### PR TITLE
Cargo Order Logging

### PR DESCRIPTION
**Pull request by Arrow768**
Adds logging of cargo orders to the database.

TODO:

- [x] Add character ids to cargo_order
- [x] Testing

### CODE REVIEW

**Comment by Arrow768** [comment]
!review (Yes, I know that travis doesnt like the migration atm)

**Review by mikomyazaki — APPROVED** [review]

File: code/controllers/subsystems/cargo.dm [hunk 1]
```diff
@@ -706,3 +709,75 @@ var/datum/controller/subsystem/cargo/SScargo
...
+	var/DBQuery/dump_query = dbcon.NewQuery("INSERT INTO `ss13_cargo_orderlog` (`game_id`, `order_id`, `status`, `price`, `ordered_by_id`, `ordered_by`, `authorized_by_id`, `authorized_by`, `received_by_id`, `received_by`, `paid_by_id`, `paid_by`, `time_submitted`, `time_approved`, `time_shipped`, `time_delivered`, `time_paid`, `reason`) \
+	VALUES (:game_id:, :order_id:, :status:, :price:, :ordered_by_id:, :ordered_by:, :authorized_by_id:, :authorized_by:, :received_by_id:, :received_by:, :paid_by_id:, :paid_by:, :time_submitted:, :time_approved:, :time_shipped:, :time_delivered:, :time_paid:, :reason:)")
+	var/DBQuery/dump_item_query = dbcon.NewQuery("INSERT INTO `ss13_cargo_orderlog_items` (`cargo_orderlog_id`, `cargo_item_id`, `amount`) \
+	VALUES (:cargo_orderlog_id:, :cargo_item_id:, :amount:)")
+	var/DBQuery/log_id = dbcon.NewQuery("SELECT LAST_INSERT_ID() AS log_id")
+	for(var/datum/cargo_order/co in all_orders)
+		//Iterate over the items in the order and build the a list with the item count
+		var/list/itemcount = list()
+		for(var/datum/cargo_order_item/coi in co.items)
+			if("[coi.ci.id]" in itemcount)
```
Comment (skull132) [inline code/controllers/subsystems/cargo.dm:737]:
```suggestion
			if(!isnull(itemcount["[coi.ci.id]"]))
```
technically faster.

File: code/modules/modular_computers/file_system/programs/civilian/cargo_control.dm [hunk 2]
```diff
@@ -108,6 +104,14 @@
 		ui.set_auto_update(TRUE)
 
 /datum/nano_module/program/civilian/cargocontrol/Topic(href, href_list)
+	if(..())
+		return TRUE
+
+	var/mob/user = usr
+	if(!istype(user))
```
Comment (skull132) [inline code/modules/modular_computers/file_system/programs/civilian/cargo_control.dm:111]: `usr` is _always_ of type `mob`. The variable `user` is pointless and should probably be avoided.

File: code/modules/modular_computers/file_system/programs/civilian/cargo_delivery.dm [hunk 3]
```diff
@@ -61,6 +59,10 @@
 	if(..())
 		return TRUE
 
+	var/mob/user = usr
+	if(!istype(user))
+		return
+	var/obj/item/card/id/I = user.GetIdCard()
```
Comment (skull132) [inline code/modules/modular_computers/file_system/programs/civilian/cargo_delivery.dm:65]: same here. You can just use `usr`.

File: code/modules/modular_computers/file_system/programs/civilian/cargo_order.dm [hunk 4]
```diff
@@ -92,12 +84,18 @@
 	if(..())
 		return TRUE
 
+	var/mob/user = usr
+	if(!istype(user))
+		return
+
+	var/obj/item/card/id/I = user.GetIdCard()
```
Comment (skull132) [inline code/modules/modular_computers/file_system/programs/civilian/cargo_order.dm:91]: And here.

File: code/modules/shuttles/shuttle_console.dm [hunk 5]
```diff
@@ -101,9 +101,13 @@
...
 		ui.set_auto_update(1)
 
-/obj/machinery/computer/shuttle_control/Topic(user, href_list)
+/obj/machinery/computer/shuttle_control/Topic(href_list, href_list)
 	..()
 
+	var/mob/user = usr
+	if(!istype(user))
+		return
+
```
Comment (skull132) [inline code/modules/shuttles/shuttle_console.dm:110]: This check is pointless.

**Review by skull132 — CHANGES_REQUESTED** [review]

**Pull request merged**